Elasticsearch json logging (#36833)
elastic/elasticsearch: docs/reference/setup/logging-config.asciidoc

appender.rolling.type = RollingFile <1>
appender.rolling.name = rolling
appender.rolling.fileName = ${sys:es.logs.base_path}${sys:file.separator}${sys:es.logs.cluster_name}.log <2>
appender.rolling.layout.type = ESJsonLayout <3>
appender.rolling.layout.type_name = server <4>

> appender.rolling.filePattern = ${sys:es.logs.base_path}${sys:file.separator}${sys:es.logs.cluster_name}-%d{yyyy-MM-dd}-%i.log.gz <5>

Reviewer: According to our new default config this should include the `_server` suffix. I think you should also align the example file name in the paragraph above with the file name here, i.e. in the end the name should be `/var/log/elasticsearch/production_server.log`.

appender.rolling.policies.type = Policies
appender.rolling.policies.time.type = TimeBasedTriggeringPolicy <6>
appender.rolling.policies.time.interval = 1 <7>
appender.rolling.policies.time.modulate = true <8>
appender.rolling.policies.size.type = SizeBasedTriggeringPolicy <9>
appender.rolling.policies.size.size = 256MB <10>
appender.rolling.strategy.type = DefaultRolloverStrategy
appender.rolling.strategy.fileIndex = nomax
appender.rolling.strategy.action.type = Delete <11>
appender.rolling.strategy.action.basepath = ${sys:es.logs.base_path}
appender.rolling.strategy.action.condition.type = IfFileName <12>
appender.rolling.strategy.action.condition.glob = ${sys:es.logs.cluster_name}-* <13>
appender.rolling.strategy.action.condition.nested_condition.type = IfAccumulatedFileSize <14>
appender.rolling.strategy.action.condition.nested_condition.exceeds = 2GB <15>
--------------------------------------------------

<1> Configure the `RollingFile` appender
<2> Log to `/var/log/elasticsearch/production.log`

> <3> Use Json layout.

Reviewer: JSON is an acronym. Hence it is "JSON" not "Json".

> <4> `type_name` is a flag populating `type` field in a `ESJsonLayout`. It is used to distinguish streams of logs when parsing.

Reviewer: Nit: "populating *the* `type` field". 

How about "It is used to distinguish streams of logs when parsing." -> "It can be used to distinguish different types of logs more easily when parsing them."?

> <5> Roll logs to `/var/log/elasticsearch/production-yyyy-MM-dd-i.log`; logs

Reviewer: According to our new default config this should include the _server suffix.

You can disable it in the `config/log4j2.properties` file by setting the deprecation
log level to `error`.

[float]
[[json-logging]]

> === Json log format

Reviewer: Json -> JSON

> To make parsing elasticsearch logs easier, logs are now printed in a json format.

Reviewer: elasticsearch -> Elasticsearch
json -> JSON

> This is configured by a log4j layout property `appender.rolling.layout.type = ESJsonLayout`.

Reviewer: log4j -> Log4J

> This layout requires a type_name attribute to be set which is used to distinguish

Reviewer: nit: enclose type_name in backticks.

logs streams when parsing.
[source,properties]
--------------------------------------------------
appender.rolling.layout.type = ESJsonLayout
appender.rolling.layout.type_name = server
--------------------------------------------------

> Each line contains a single json with set of attributes that are configured in `ESJsonLayout`.

Reviewer: ```suggestion
Each line contains a single JSON document with the properties configured in `ESJsonLayout`.
```

> See this class javadoc for more details.

Reviewer: I am not sure we should point users of our reference docs to the Javadoc. In the end it would mean that they'd either need to checkout the sources or go to Github, find the right class and read the Javadocs which seems inconvenient. I wonder whether we should go the extra mile and document the properties here as well although the information is duplicated then. Or we don't document it in Javadoc and only in the reference docs. Either way, I think we can improve this

Author: or maybe I could point them directly?
We do in sam documentation pages already. For isntance: https://www.elastic.co/guide/en/elasticsearch/client/java-rest/6.0/java-rest-high-query-builders.html

> When the log event contains a `Throwable`, json will be formatted multiline where first line contains

Reviewer: I think this too technical. How about: "Exception stack traces will be formatted as multiple lines ..."

Author: `However if JSON documents contains exception, stack trace will be formatted as a multiline.
`

Author: will reword

> all the regular attributes and subsequent lines will contain stacktrace formatted as a json array.

Reviewer: I think the term to use is "properties", not "attributes" but I might be wrong. Also: json -> JSON

> NOTE: You can still use your own custom layout. To do that just replace line

Reviewer: Scratch "just"? It does not convey any meaning here IMHO? Also: "replace line" -> replace *the* line"

> `appender.rolling.layout.type` with the layout of your preference. See sample below:

Reviewer: "with the layout of your preference" -> "with a different layout" or "with your preferred layout"?

[source,properties]
--------------------------------------------------
appender.rolling.type = RollingFile
appender.rolling.name = rolling

> appender.rolling.fileName = ${sys:es.logs.base_path}${sys:file.separator}${sys:es.logs.cluster_name}.log

Reviewer: Add `_server` suffix for consistency?

appender.rolling.layout.type = PatternLayout
appender.rolling.layout.pattern = [%d{ISO8601}][%-5p][%-25c{1.}] [%node_name]%marker %.-10000m%n

> appender.rolling.filePattern = ${sys:es.logs.base_path}${sys:file.separator}${sys:es.logs.cluster_name}-%d{yyyy-MM-dd}-%i.log.gz

Reviewer: Add `_server` suffix for consistency?

Author: as discussed not adding to archive names

You can disable it in the `config/log4j2.properties` file by setting the deprecation
log level to `error`.

[float]
[[json-logging]]
=== JSON log format

To make parsing Elasticsearch logs easier, logs are now printed in a JSON format.
This is configured by a Log4J layout property `appender.rolling.layout.type = ESJsonLayout`.
This layout requires a `type_name` attribute to be set which is used to distinguish
logs streams when parsing.
[source,properties]
--------------------------------------------------
appender.rolling.layout.type = ESJsonLayout
appender.rolling.layout.type_name = server
--------------------------------------------------
:es-json-layout-java-doc: {elasticsearch-javadoc}/org/elasticsearch/common/logging/ESJsonLayout.html

Each line contains a single JSON document with the properties configured in `ESJsonLayout`.
See this class {es-json-layout-java-doc}[javadoc] for more details.

> However if a JSON document contains exception, it will be formatted as a multiline.

Reviewer: Nits: 
* "contains **an** exception"
* "formatted as a multiline" -> "printed over multiple lines"

> First line will contain regular properties and subsequent lines will contain

Reviewer: Nit: "**The** first line"

> stacktrace formatted as a JSON array.

Reviewer: Nit: "**the** stacktrace"

> NOTE: You can still use your own custom layout. To do that replace the line

Reviewer: I think the docs need to take into account that we now log as JSON *and* in plain text.

Each line contains a single JSON document with the properties configured in `ESJsonLayout`.
See this class {es-json-layout-java-doc}[javadoc] for more details.

> However if a JSON document contains an exception, it will be printted over multiple lines.

Reviewer: typo: "printted" -> "printed"
